include/closure_hydrator.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <numeric>
+#include <optional>
+#include <vector>
+
+#include "log.h"
+#include "ranges.h"
+#include "sycl_wrappers.h"
+#include "types.h"
+
+namespace celerity::detail {
+
+// To avoid additional register pressure, we embed hydration IDs into pointers for
+// accessors, with the assumption that a real pointer will never be in the
+// range [0, max_hydration_id]. Embedding / extracting are currently no-ops
+// and the associated helper functions only exist for documentation purposes.
+// This number puts an effective limit on the number of task objects (accessors
+// etc.) that can be captured into a command function.
+constexpr size_t max_hydration_id = 128;
+
+template <typename T>
+constexpr bool can_embed_hydration_id = std::bool_constant<sizeof(hydration_id) == sizeof(T)>::value;
+
+template <typename T>
+T embed_hydration_id(const hydration_id hid) {
+	static_assert(can_embed_hydration_id<T>);
+	assert(hid > 0); // Has to be greater than zero so nullptr is not considered an embedded id
+	assert(hid <= max_hydration_id);
+	T result;
+	std::memcpy(&result, &hid, sizeof(hid));
+	return result;
+}
+
+template <typename T>
+hydration_id extract_hydration_id(const T value) {
+	static_assert(can_embed_hydration_id<T>);
+	hydration_id result;
+	std::memcpy(&result, &value, sizeof(value));
+	return result;
+}
+
+template <typename T>
+bool is_embedded_hydration_id(const T value) {
+	static_assert(can_embed_hydration_id<T>);
+	const auto hid = extract_hydration_id(value);
+	return hid > 0 && hid <= max_hydration_id;
+}
+
+/**
+ * The closure hydrator is used to inject information into objects (currently host/device and local accessors) that have been captured into a lambda closure.
+ * We abuse the copy constructor of the captured objects to modify them while the containing closure is being copied by the hydrate() function.
+ * Accessors request their corresponding information by means of per-closure unique "hydration ids" that are assigned upon accessor creation.
+ *
+ * The hydrator is implemented as a thread-local singleton that needs to be explicitly enabled per-thread. This is because kernel command function
+ * closures may be copied any number of times after having been passed to SYCL, which should not trigger the hydration mechanism.
+ */
+class closure_hydrator {
+  public:
+	struct accessor_info {
+		void* ptr;
+		range<3> buffer_range;
```

Comment:
Last-minute request, can we improve the naming of these members to clarify which ones are in virtual-buffer and which are in backing-buffer / allocation coordinates?

Reply:
Done, I've gone ahead and clarified the terminology also for accessors.